src/main/java/com/blamejared/compat/tcomplement/highoven/MixRecipeManager.java

```diff
@@ -0,0 +1,130 @@
+		this.init = false;
+		this.input = InputHelper.toFluid(input);
+		this.output = InputHelper.toFluid(output);
+		this.additives = new LinkedList<>();
+		this.removedAdditives = new HashMap<>(MixAdditive.values().length, 1.0f);
+		for (MixAdditive type : MixAdditive.values()) {
+			this.removedAdditives.put(type, new LinkedList<>());
+		}
+	}
+
+	@ZenMethod
+	public MixRecipeManager addOxidizer(@NotNull IIngredient oxidizer, int consumeChance) {
+		this.additives.add(new SimpleEntry<>(new RecipeMatchIIngredient(oxidizer, consumeChance), MixAdditive.OXIDIZER));
+		return this;
+	}
+
+	@ZenMethod
+	public MixRecipeManager addReducer(@NotNull IIngredient reducer, int consumeChance) {
+		this.additives.add(new SimpleEntry<>(new RecipeMatchIIngredient(reducer, consumeChance), MixAdditive.REDUCER));
+		return this;
+	}
+
+	@ZenMethod
+	public MixRecipeManager addPurifier(@NotNull IIngredient purifier, int consumeChance) {
+		this.additives.add(new SimpleEntry<>(new RecipeMatchIIngredient(purifier, consumeChance), MixAdditive.PURIFIER));
+		return this;
+	}
+
+	@ZenMethod
+	public MixRecipeManager removeOxidizer(IIngredient oxidizer) {
+		if (oxidizer != null && removedAdditives.get(MixAdditive.OXIDIZER) != null) {
+			removedAdditives.get(MixAdditive.OXIDIZER).add(new RecipeMatchIIngredient(oxidizer));
+		} else {
+			removedAdditives.put(MixAdditive.OXIDIZER, null);
+		}
+		return this;
+	}
+
+	@ZenMethod
+	public MixRecipeManager removeReducer(IIngredient reducer) {
+		if (reducer != null && removedAdditives.get(MixAdditive.REDUCER) != null) {
+			removedAdditives.get(MixAdditive.REDUCER).add(new RecipeMatchIIngredient(reducer));
+		} else {
+			removedAdditives.put(MixAdditive.REDUCER, null);
+		}
+		return this;
+	}
+
+	@ZenMethod
+	public MixRecipeManager removePurifier(IIngredient purifier) {
+		if (purifier != null && removedAdditives.get(MixAdditive.PURIFIER) != null) {
+			removedAdditives.get(MixAdditive.PURIFIER).add(new RecipeMatchIIngredient(purifier));
+		} else {
+			removedAdditives.put(MixAdditive.PURIFIER, null);
+		}
+		return this;
+	}
+	
+	@ZenMethod
+	public void register() {
+		if (!this.init) {
+			MinecraftForge.EVENT_BUS.register(this);
+			init = true;
+		}
+	}
+
+	@SubscribeEvent
+	public void onTinkerRegister(TCompRegisterEvent.HighOvenMixRegisterEvent event) {
+		if (event.getRecipe().matches(this.input, this.output)) {
+			this.additives.forEach((SimpleEntry<RecipeMatch, MixAdditive> entry) -> event.getRecipe()
+					.addAdditive(entry.getValue(), entry.getKey()));
+		}
+	}
+
+	@SubscribeEvent
+	public void onTinkerRegister(TCompRegisterEvent.HighOvenMixAdditiveEvent event) {
+		if (event.getRecipe().matches(input, output)) {
+			List<RecipeMatch> removals = removedAdditives.get(event.getType());
+			if (removals != null) {
+				for (RecipeMatch removal : removals) {
+					//TODO: Fix THIS!
+					if (removal.matches(NonNullList.from(event.getAdditive().getInputs())).isPresent()) {
```

Comment:
This is not going to work well, `getInputs()` gets you a list of every possible match for things such as oredict, but the matches list method receiving a list is intended as multiple inventory slots. What you really want here is iterate the list of additives in the recipe and see if any match the ingredient

That would look something like
```
for (ItemStack stack : recipe.getAdditive().getInputs())
  if (removal.matches(stack)) {
    event.setCanceled(true);
    break;
  }
```

Reply:
Thanks ! I really misunderstood what `RecipeMatch.matches()` does. This should work properly now.